NEAR Simulator -> workspaces-rs migration guide (#99)
near/sdk-docs: docs/testing/workspaces-migration-guide.md

---
sidebar_position: 3
sidebar_label: Workspaces Migration Guide
title: "Workspaces Migration Guide"
---

# Workspaces Migration Guide

NEAR Simulator was meant to be an in-place replacement of a blockchain environment for the purpose of testing NEAR contracts. However, simulating NEAR ledger turned out to be a much more complex endeavour than was anticipated. Eventually, the idea of workspaces was born - a library for automating workflows and writing tests for NEAR smart contracts using a real NEAR network (localnet, testnet or mainnet). Thus, NEAR Simulator is being deprecated in favor of [`workspaces-rs`](https://github.com/near/workspaces-rs), the Rust edition of workspaces. As the two libraries have two vastly different APIs this guide was created to ease the migration process for developers.

:::danger

> TODO: I do not have a whole lot of context here why exactly simtests were not suitable for our purposes, so if anyone wants to elaborate the preceding paragraph please do.

Author: Can someone help filling this gap? @ChaoticTempest you probably have the most amount of context here, feel free to write it as a comment or push a commit to my branch.

Reviewer: A few reasons I can remember:
- near-sdk-sim was hooking into parts of nearcore that were not meant to be released, in the most recent version those crates aren't released so sim is currently using duplicate deps (maintenance nightmare)
- Not a fully accurate simulation because it just used a subset of the runtime in a specific way -- can't rely on this
- Bulky dependencies pulled in (drastically increases compile time)
- Unergonomic API, not specific to this strategy, but likely would have had to be re-built
- Can't test parallel transactions easily -- current pattern would process blocks until a transaction succeeded but you can't create specific conditions, which is required for a strategy like this that isn't fully simulated

I'll let Phuong share opinions when he's back since he likely will have some inputs for this

Reviewer 2: > Not a fully accurate simulation because it just used a subset of the runtime in a specific way -- can't rely on this

And thus couldn't measure gas burnt accurately. Also, all the intricacies of nearcore (like protocol features) wouldn't be one-to-one with the runtime since the runtime was just code built on top of VM logic. 

People would also need to write their own automation scripts to deploy to testnet, so we'd end up with very split workflows for testing.

:::

> This guide presumes that you are transitioning from near-sdk-sim `3.2.0` (the last non-deprecated release) to `workspaces-rs` `0.2.1`. Given that near-sdk-sim is deprecated, it is very unlikely that its API is going to ever change, but future releases of `workspaces-rs` might. Hopefully this guide is going to be helpful even if you are migrating your project to a more recent version, but also feel free to migrate your tests to `0.2.1` using this guide first and upgrade to the most recent workspaces-rs version later by looking at the release notes to see how public API has changed since `0.2.1`.

Reviewer 2: nit: this paragraph is a bit wordy. Maybe something like this would be better

```suggestion
This guide presumes that you are transitioning from near-sdk-sim `3.2.0` (the last non-deprecated release) to `workspaces-rs` `0.2.1`. Given that near-sdk-sim is deprecated, it is very unlikely that its API will ever change, but future releases of `workspaces-rs` might. Hopefully, this guide will be helpful even if you are migrating your project to a more recent workspaces version. If workspaces have changed, feel free to migrate your tests to `0.2.1` first using this guide and upgrade to the most recent workspaces-rs version later by looking at the release notes to see how public API has changed since `0.2.1`.
```

Author: Yeah, that is much nicer

* `testnet` - an environment much closer to the real world; you can test integrations with other deployed contracts on testnet without bearing any financial risk.
* `mainnet` - a network with reduced amount of features due to how dangerous it can be to do transactions there, but can still be useful for automating deployments and pulling deployed contracts.

In this guide we will be focusing on `sandbox` since it covers the same use cases NEAR Simulator did. But of course feel free to explore whether other networks can be of potential use to you when writing new tests/workflows.

One of the ways to initialize simulator and deploy a contract is shown below (the other way is through `deploy!` macro which we will look at in the next section):

```rust title="Deployment - near-sdk-sim"
use near_sdk_sim::{init_simulator, to_yocto};

near_sdk_sim::lazy_static_include::lazy_static_include_bytes! {
    WASM_BYTES => "res/contract.wasm",
}

const ID: &str = "contract-id";

...

let root = init_simulator(...);
let contract = root.deploy(&WASM_BYTES, ID.parse().unwrap(), to_yocto("5"));
```

> Although `workspaces-rs` provides a way to specify the account id for a contract to be deployed, usually it does not matter in the context of a single test. If you are fine with generating a random developer account and initializing it with 100N, then you can use replace the snippet above with this:

Reviewer 2: ```suggestion
Although `workspaces-rs` provides a way to specify the account id for a contract to be deployed, usually it does not matter in the context of a single test. If you are fine with generating a random developer account and initializing it with 100N, then you can replace the snippet above with this:
```

```rust title="Deployment - workspaces-rs"
let worker = workspaces::sandbox().await?;
let contract = worker.dev_deploy(include_bytes!("../res/contract.wasm")).await?;
```

Alternatively, use this if you care about the account id:

```rust title="Deployment - workspaces-rs (with explicit account id)"
let worker = workspaces::sandbox().await?;
let (_, sk) = worker.dev_generate().await;
let id: AccountId = "contract-id".parse()?;
let contract = worker
    .create_tla_and_deploy(
        id,
        sk,
        include_bytes!("../examples/res/non_fungible_token.wasm"),
    )
    .await?
    .result;
```

Or, if you want to create a subaccount with a certain balance:

```rust title="Deployment - workspaces-rs (with explicit balance)"
use near_units::parse_near;

let worker = workspaces::sandbox().await?;
let id: AccountId = "contract-id".parse()?;
let contract = worker
    .root_account()
    .create_subaccount(&worker, id)
    .initial_balance(parse_near!("5 N"))
    .transact()
    .await?

>     .result;

Reviewer 2: This doesn't actually deploy the contract, since this only creates the subaccount. Also, better to use `into_result()?` instead of `.result`

```

:::danger

> TODO: Is there a reason why we can't control the initial balance with `dev_deploy`?

Author: @ChaoticTempest question to you

Reviewer 2: `dev_deploy` can't supply the initial balance since testnet controls this amount in the helper contract which is what we're using to create dev accounts in testnet. So, to make it simple, we don't supply it at all for sandbox even. If someone wants to supply an amount in sandbox, they can grab the root account and do:
```rust
let root = worker.root_acount();
root.create_subaccount(...)
   .initial_balance(...)
   ...
```

Workspaces have a unified way of making all types of calls via a [builder](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html) pattern. Generally, calls are constructed by following these steps:

1. Create a `CallBuilder` by invoking `Contract::call`
2. Pass function call arguments via `CallBuilder::args_json` or `CallBuilder::args_borsh` depending on which serialization algorithm your contract is using
3. Configure gas and deposit (if needed) via `CallBuilder::gas` and `CallBuilder::deposit`
4. Finalize the call by consuming builder via `CallBuilder::transaction` or `CallBuilder::view` depending on what kind of call you want to make

Reference this migration of Example 1 for migrating your own calls:

```rust title="Calls - workspaces-rs"
contract
    .call(&worker, "ft_transfer")
    .args_json((alice.id(), transfer_amount, Option::<bool>::None))?
    .gas(300_000_000_000_000)
    .deposit(ONE_YOCTO)
    .transact()
    .await?;

let root_balance: U128 = contract
    .call(&worker, "ft_balance_of")
    .args_json((contract.id(),))?
    .view()
    .await?
    .json()?;
```

:::note

> Note that you have to pass arguments as any serializable type representing a sequential list. Tuples are usually the best candidate due to their heterogeneous nature (remember that you can construct a unary tuple by placing a comma before the closing bracket like this: `(el,)`).

Reviewer 2: Might want to also mention the `json!()` macro is acceptable here as well

    .function_call(
        "ft_transfer_call".to_string(),
        json!({
            "receiver_id": defi_contract.account_id(),
            "amount": transfer_amount.to_string(),
            "msg": "10",
        })
        .to_string()
        .into_bytes(),
        300_000_000_000_000 / 2,
        1,
    )
    .function_call(
        "storage_unregister".to_string(),
        json!({
            "force": true
        })
        .to_string()
        .into_bytes(),
        300_000_000_000_000 / 2,
        1,
    )
    .submit();
```

There are no caveats here, the snippet can be straightforwardly mapped into the following:

```rust title="Batch Transaction - workspace-rs"
let res = contract
    .batch(&worker)
    .call(
        Function::new("ft_transfer_call")
            .args_json((defi_contract.id(), transfer_amount, Option::<String>::None, "10"))?
            .gas(300_000_000_000_000 / 2)
            .deposit(1),
    )
    .call(
        Function::new("storage_unregister")
            .args_json((Some(true),))?
            .gas(300_000_000_000_000 / 2)
            .deposit(1),
    )
    .transact()
    .await?;
```

## Inspecting logs

:::danger

> TODO: I found some potential inconsistencies in what I wrote about the log inspection behavior, need some time to look at this further

Reviewer 2: what kind of inconsistencies did you find. The section you wrote afterwards describes it well

);
```

The `workspaces-rs` counterpart might seem almost identical at the first look:

```rust title="Logs - workspaces-rs"
assert_eq!(
    res.logs()[1],
    format!("Closed @{} with {}", contract.id(), initial_balance.0 - transfer_amount.0)
);
```

However, it can actually behave differently depending on your use case, because while near-sdk-sim version only returns the logs from the transaction, the workspaces version returns all logs from both the transaction and receipt outcomes. If you want a literal counterpart, please use `res.outcome().logs`.

Another common use case is examining receipt outcome logs like this:

```rust title="Logs - nead-sdk-sim"
let outcome = res.get_receipt_results().remove(5).unwrap();

assert_eq!(outcome.logs()[0], "The account of the sender was deleted");
assert_eq!(
    outcome.logs()[2],
    format!("Account @{} burned {}", root.account_id(), 10)
);
```

Which is straightforwardly replaced with:

```rust title="Logs - workspaces-rs"
let outcome = &res.receipt_outcomes()[5];
assert_eq!(outcome.logs[0], "The account of the sender was deleted");
assert_eq!(outcome.logs[2], format!("Account @{} burned {}", contract.id(), 10));
```

## Profiling gas

> NEAR Simulator has not been providing realistic gas estimates (TODO: needs elaboration - since when and why?), so the `workspaces-rs` gas estimates might come as a surprise. The good news is that they do reflect reality!

Author: @ChaoticTempest also probably yours :)

Reviewer 2: > TODO: needs elaboration - since when and why?

it never had accurate gas estimations since it only tried to mirror nearcore, but nearcore has extra things on top which consumes gas (like cross-contract calls are processed separately from the same transaction and that incurs gas fees)

I'd give a warning that since workspaces-rs is now using accurate gas measurements, some contracts that were previously being tested with sdk-sim might not work the same, and that they should do their due diligence if they plan to deploy to mainnet. I mean, either ways, if they tried deploying the same contracts with infeasible gas estimates to testnet, it would have never worked. So we can say that workspaces offers the better experience here and aligns very well with what you can do on testnet and mainnet
